### PR TITLE
Add possibility to decode generated encoded root token to api

### PR DESCRIPTION
**Pull request by l-with**
solves #5236 (closed)

this makes it possible to get a root token completely using the API by adding the possibility to decode the generated encoded root token at /v1/sys/decode-token

I think it's a good thing if the API supports the same things as the cli (here vault itself).

### CODE REVIEW

**Comment by ncabatoff** [comment]
Hi @l-with, thanks for writing this improvement.

I can't remember why we didn't do this already.  Maybe we just thought that it wasn't worth creating an endpoint to do an XOR?  If so that was probably wrongheaded, this seems like a good addition to me today.

My main objection to the way you've implemented this is that you're following the example of other code related to root token generation and putting the handler in the http package, then adding an explicit mux.Handle call to the handler.  I would like us to have as few such handlers as possible.  The ones we have currently all have some reason why they can't be treated like regular logical requests, e.g. they don't want to return a JSON response, or they want to use unseal/recovery keys as their authentication mechanism, or they want to work even when Vault is sealed.

I can't see of any reason why your new endpoint couldn't be handled as a logical request, as in a new SystemBackend method invoked via a new path handler added in NewSystemBackend.  I'm guessing you want to make it Unauthenticated, which is probably fine, and which you can also configure in NewSystemBackend.

**Comment by l-with** [comment]
@ncabatoff to avoid double work, my first question:
what is your suggestion for the API endpoint?

I would use generate-root/decode-token, what do you think?

**Comment by l-with** [comment]
Hi @ncabatoff 
I started the refactoring. I would be appreciated to get your feedback, if this is the way what you suggested.

**Comment by ncabatoff** [comment]
> Hi @ncabatoff I started the refactoring. I would be appreciated to get your feedback, if this is the way what you suggested.

Yup, this is exactly what I had in mind.  I think you're almost there, except for the tests.

**Comment by l-with** [comment]
> > Hi @ncabatoff I started the refactoring. I would be appreciated to get your feedback, if this is the way what you suggested.
> 
> Yup, this is exactly what I had in mind. I think you're almost there, except for the tests.

Thanks, I'll refactor the test also.
I have a question: Can I avoid returning the complete type Response or can I extract the data attribute?

**Comment by l-with** [comment]
@ncabatoff 
I implemented the test in logical_system_test.go.
Please tell me if this is the right place for testing the new logical backend.

File: vault/logical_system_test.go [hunk 1]
```diff
@@ -2420,6 +2420,62 @@ func TestSystemBackend_enableAudit(t *testing.T) {
...
+		)
+	}
+}
+
+func TestSystemBackend_decodeToken(t *testing.T) {
+	encodedToken := "Bxg9JQQqOCNKBRICNwMIRzo2J3cWCBRi"
+	otp := "3JhHkONiyiaNYj14nnD9xZQS"
+	tokenExpected := "4RUmoevJ3lsLni9sTXcNnRE1"
+
+	t.Run("basic", func(t *testing.T) {
```
Comment (ncabatoff) [inline vault/logical_system_test.go:2446]: I don't object to the test as written, but since there's no side effect from the new endpoint, the test could be written to share the backend/core returned by testCoreSystemBackend across all four subtests, rather than creating a fresh one for each.  This would be a little more performant and conventional, but I don't mind if you choose not to do this, it's more of a suggestion than a request.

File: vault/logical_system_test.go [hunk 2]
```diff
@@ -2420,6 +2420,62 @@ func TestSystemBackend_enableAudit(t *testing.T) {
...
 }
 
+func missingPayloadTest(path string, data map[string]interface{}) func(t *testing.T) {
+	return func(t *testing.T) {
+		_, b, _ := testCoreSystemBackend(t)
+		req := logical.TestRequest(t, logical.UpdateOperation, path)
+		req.Data = data
+		resp, err := b.HandleRequest(namespace.RootContext(nil), req)
+		if err == nil {
+			t.Fatalf("err: %v", err)
```
Comment (ncabatoff) [inline vault/logical_system_test.go:2430]: We know err is nil here, so maybe a simple literal string like "missing payload" makes more sense as the fatal output than "err: nil".

File: api/go.mod [hunk 3]
```diff
@@ -19,6 +19,7 @@ require (
 	github.com/hashicorp/go-secure-stdlib/parseutil v0.1.6
 	github.com/hashicorp/go-secure-stdlib/strutil v0.1.2
 	github.com/hashicorp/hcl v1.0.0
+	github.com/hashicorp/vault/sdk v0.9.0
```
Comment (ncabatoff) [inline api/go.mod:22]: I don't see why api would need to change its dependencies for your PR, and also this is undoing some of my recent work that ensured that api doesn't depend on sdk; can you either revert this change or explain to me why it's necessary please?

**Review by ncabatoff — COMMENTED** [review]
A couple of little cleanups are needed, but overall this looks nearly done

File: vault/logical_system_paths.go [hunk 4]
```diff
@@ -466,6 +466,33 @@ func (b *SystemBackend) configPaths() []*framework.Path {
...
+			},
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.UpdateOperation: &framework.PathOperation{
+					Callback: b.handleGenerateRootDecodeTokenUpdate,
+					DisplayAttrs: &framework.DisplayAttributes{
+						OperationVerb:   "decode",
+						OperationSuffix: "encoded-token",
+					},
+					Summary: "Decodes the encoded token with the otp.",
+					Responses: map[int][]framework.Response{
```
[the remark this answers is not in the document]
Comment (ncabatoff) [reply vault/logical_system_paths.go:488]: Actually I'm probably mistaken about that, please disregard.

File: vault/logical_system_paths.go [hunk 5]
```diff
@@ -503,7 +530,7 @@ func (b *SystemBackend) configPaths() []*framework.Path {
 				},
 			},
 			Operations: map[logical.Operation]framework.OperationHandler{
-				logical.ReadOperation: &framework.PathOperation{
+				logical.UpdateOperation: &framework.PathOperation{
```
[the remark this answers is not in the document]
Comment (l-with) [reply vault/logical_system_paths.go]: ups, sorry I fixed it

File: vault/logical_system_test.go (hunk 1 above)
Comment (l-with) [reply vault/logical_system_test.go]: I liked to try out returning a function by a function. I understand the performance suggestion and refactored the test.

File: vault/logical_system_test.go (hunk 2 above)
Comment (l-with) [reply vault/logical_system_test.go]: OK, I decided to return a message explaining the error.

File: api/go.mod (hunk 3 above)
Comment (l-with) [reply api/go.mod]: Surely unnecessary. I think this was the result of an intermediate step.
I removed it.

Comment (l-with) [reply api/go.mod]: I additionally also executed go mod tidy.

**Review by ncabatoff — APPROVED** [review]
Looks good!  One last minor improvement suggestion for you to consider, it's optional though.

**Review by ncabatoff — APPROVED** [review]
Looks good!  One last minor improvement suggestion for you to consider, it's optional though.

File: vault/logical_system_test.go [hunk 6]
```diff
@@ -2420,6 +2420,58 @@ func TestSystemBackend_enableAudit(t *testing.T) {
 	}
 }
 
+func TestSystemBackend_decodeToken(t *testing.T) {
+	encodedToken := "Bxg9JQQqOCNKBRICNwMIRzo2J3cWCBRi"
```
[the remark this answers is not in the document]
Comment (ncabatoff) [reply vault/logical_system_test.go:2424]: Ah, don't worry about it in that case, let's not expand the scope of this PR needlessly.  Thanks for the extra details.

**Comment by ncabatoff** [comment]
Oh, sorry, I forgot about docs.  Could I trouble you to add the new endpoint to the api docs?  I can give you some pointers if it's not clear what's involved... you seem to be pretty good at figuring this stuff out on your own though so I'm not sure you need the help.

**Comment by l-with** [comment]
> Oh, sorry, I forgot about docs. Could I trouble you to add the new endpoint to the api docs? I can give you some pointers if it's not clear what's involved... you seem to be pretty good at figuring this stuff out on your own though so I'm not sure you need the help.

Do you mean openAPI doc or is https://developer.hashicorp.com/vault/api-docs or parts from generated by the repository?

**Comment by ncabatoff** [comment]
> > Oh, sorry, I forgot about docs. Could I trouble you to add the new endpoint to the api docs? I can give you some pointers if it's not clear what's involved... you seem to be pretty good at figuring this stuff out on your own though so I'm not sure you need the help.
> 
> Do you mean openAPI doc or is https://developer.hashicorp.com/vault/api-docs or parts from it generated by the repository?

The latter, e.g. https://github.com/hashicorp/vault/blob/main/website/content/api-docs/system/generate-root.mdx
I suspect this will mean a new mdx file in that directory (since I don't see an existing one that this would fit into), and a small addition to https://github.com/hashicorp/vault/blob/main/website/data/api-docs-nav-data.json to add the new path to the sidebar.

**Comment by l-with** [comment]
> > > Oh, sorry, I forgot about docs. Could I trouble you to add the new endpoint to the api docs? I can give you some pointers if it's not clear what's involved... you seem to be pretty good at figuring this stuff out on your own though so I'm not sure you need the help.
> > 
> > 
> > Do you mean openAPI doc or is https://developer.hashicorp.com/vault/api-docs or parts from it generated by the repository?
> 
> The latter, e.g. https://github.com/hashicorp/vault/blob/main/website/content/api-docs/system/generate-root.mdx I suspect this will mean a new mdx file in that directory (since I don't see an existing one that this would fit into), and a small addition to https://github.com/hashicorp/vault/blob/main/website/data/api-docs-nav-data.json to add the new path to the sidebar.

Sorry, I deleted my comment, since I already started exactly what you suggested.

File: changelog/20595.txt [hunk 7]
```diff
@@ -0,0 +1,3 @@
+```release-note:improvement
+Add possibility to decode a generated encoded root token via the rest API
```
Comment (ncabatoff) [inline changelog/20595.txt:2]:
```suggestion
core: Add possibility to decode a generated encoded root token via the rest API
```

**Review by ncabatoff — COMMENTED** [review]
Docs look good, one last tweak to the changelog and I'll merge it

File: changelog/20595.txt (hunk 7 above)
Comment (l-with) [reply changelog/20595.txt]: It's great to work with such a very fast answering contributor!
done

**Comment by ncabatoff** [comment]
Sorry, one more request: one of our checks is failing:

```
Error: vault/logical_system_test.go:2423:1: Test TestSystemBackend_decodeToken is missing a go doc
```

Could you add a godoc to the test func explaining its purpose please?

File: changelog/20595.txt (hunk 7 above)
Comment (l-with) [reply changelog/20595.txt]: It perfectly matched my flow!

**Comment by l-with** [comment]
> Sorry, one more request: one of our checks is failing:
> 
> ```
> Error: vault/logical_system_test.go:2423:1: Test TestSystemBackend_decodeToken is missing a go doc
> ```
> 
> Could you add a godoc to the test func explaining its purpose please?

done
